### DISCUSSION TITLE
object.refine doesnt always detect unmatching fields

### DISCUSSION THREAD

**Discussion by dev-1234567** [colinhacks/zod#2782] [Q&A] [answered]
Im using the following for password and confirm_password to check that the fields match: - 

```ts
const passwordCompoundField = z
  .object({
    password: PASSWORD,
    confirm_password: z.string().min(1, 'Required'),
  })
  .refine((data) => data.password === data.confirm_password, {
    message: "The passwords didn't match. Please try again",
    path: ['confirm_password'],
  });
```

This works fine in the usual scenario password is entered, confirmed password is entered (which doesnt match) and the error message will show under the confirm password field. However with the following scenario the error message doesnt show: - 

password: Example111
confirm_password~: Example11

Error message will show below confirm_password field

If i then amend the password field to the below: -

password: Example11
confirm_password~: Example11

The error message still remains.

To get the error to disappear i have to delete the last '1' from confirm_password then retype

**Comment by rwlanup** [comment]
@dev-1234567 I think it's your UI causing bug rather than zod. Please check your UI logic.

**Comment by JacobWeisenburger** [comment, marked as answer]
I think @rwlanup is correct:

> @dev-1234567 I think it's your UI causing bug rather than zod. Please check your UI logic.

---

Everthing seems to be working as expected:

```ts
const schema = z.object( {
    password: z.string().min( 1, 'Required' ),
    confirm_password: z.string().min( 1, 'Required' ),
} ).refine( data => data.password === data.confirm_password, {
    message: `The passwords didn't match. Please try again`,
    path: [ 'confirm_password' ],
} )

console.log(
    schema.safeParse( {
        password: 'Example11',
        confirm_password: 'Example11',
    } ).success
) // true

console.log(
    schema.safeParse( {
        password: 'Example111',
        confirm_password: 'Example11',
    } ).success
) // false
```

Please let me know if there is something I am missing.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
